Travis CI - Add support for runtime system operational test

tomereli opened the pull request:
CI support for run-time operational testing using the docker runner which runs prplmesh controller and agent in one container, then uses docker exec command to call prplmesh_utils.sh status which fails if not in operational mode, thus failing CI.

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -136,19 +140,46 @@ stop_function() {
     [ "$PRPLMESH_MODE" = "CA" -o "$PRPLMESH_MODE" = "A" ] && prplmesh_agent_stop
 }
 
+main_agent_operatioanl() {
```
typo: operatioanl -> operational

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -136,19 +140,46 @@ stop_function() {
     ps -aux | grep local_bus | grep -v grep
+
+    # check for operational status
+    if [ "$PRPLMESH_USER" != "" ]; then
```
PRPLMESH_USER is not used, and I don't see it's purpose.

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -136,19 +140,46 @@ stop_function() {
+    fi
+
+    error=false
+    main_agent_operatioanl $LOGS_PATH && success "Main agent operational" || {
```
operational again

arnout replied:
`&&` and `||` order of evaluation is not portable. Better use an explicit `if` condition.

This could also be factored in a function
```
report() {
    msg="$1"; shift
    if "$@"; then
        success " OK  $msg"
    else
        err "FAIL $msg"
        error=true
    fi
}
...
    report "Main agent operational" \
        main_agent_operational $LOGS_PATH
```

tomereli replied:
But error is local variable so how can I use it after the report? I want to dump the log on failure.

arnout replied:
shell doesn't have local variables.

bash has local variables, but only if explicitly declared with `local`.

You also get a kind of local variable by doing things in a subshell with `(...)`.

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -136,19 +140,46 @@ stop_function() {
 
 usage() {
-    echo "usage: $(basename $0) {start|stop|restart|status} [-hvsm]"
+    echo "usage: $(basename $0) {start|stop|restart|status} [-hvsmu]"
```
Not needed.

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -136,19 +140,46 @@ stop_function() {
 
 main() {
-    OPTS=`getopt -o 'hvm:s' --long verbose,help,mode:skip-platform -n 'parse-options' -- "$@"`
+    OPTS=`getopt -o 'hvm:u:s' --long verbose,help,mode:skip-platform -n 'parse-options' -- "$@"`
```
Not needed.

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -159,6 +190,7 @@ main() {
             -v | --verbose)       VERBOSE=true; shift ;;
             -h | --help)          usage; exit 0; shift ;;
             -m | --mode)          PRPLMESH_MODE="$2"; shift; shift ;;
+            -u | --user)          PRPLMESH_USER="$2"; shift; shift ;;
```
Not needed.

arnout commented on .travis.yml:
```diff
@@ -10,6 +10,10 @@ script:
   - cd tools/docker
```
type in commit message: continues -> continuous

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -136,19 +140,46 @@ stop_function() {
+    radio_agent_operational $LOGS_PATH wlan2 && success "wlan2 radio agent operational" || {
+        err "wlan2 radio agent not operational"; error=true
+    }
+    return $error
```
Should be either just
```
$error
```
(running `true` or `false`, and the last command of the function gives the exit code), or
``
$error || return 1
```

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -161,15 +161,21 @@ status_function() {
     fi
 
-    error=false
+    error=0
```
This should already have been done in the first commit.

tomereli replied:
changed back to false / true following your previous comment

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -161,15 +161,21 @@ status_function() {
 
-    error=false
+    error=0
+    cat $LOGS_PATH/beerocks_agent.log
```
This is not nice.It would be better to make sure the logs are mapped to the host and export them as build artifacts for later inspection. But that's for some other day. :ok:

tomereli replied:
This is intentional, we do not map the logs since the host is running multiple instances.

arnout commented on tools/docker/README.md:
````diff
@@ -22,7 +22,7 @@ The docker builder image can be used to build prplMesh via `maptools.py build` c
 
 ```bash
-sudo ./builder/build.sh <maptools.py build arguments>
+sudo ./build.sh <maptools.py build arguments>
````
Why this change?

tomereli replied:
Ostap's - I moved the build.sh, run.sh to the docker directory in one of the previous commits not part of this PR and forgot to change the README. the builder/runner contain now the images only.

arnout commented on tools/docker/README.md:
```diff
@@ -33,6 +33,10 @@ The runner docker allows running multiple containers, for example one with a con
 
+## Docker tester
+
+A simple wrapper which calls `sudo docker exec <container name> -it <path/to>/build/install/scripts/prplmesh_utils.sh status which prints the prplMesh operational status of the main and radio agents.
```
Missing closing backquote.

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -8,11 +8,11 @@ dbg() {
 
 err() {
-	echo -e '\033[1;31m'"$@"'\033[0m'
+    echo '\033[1;31m'"$@"'\033[0m'
```
Should be printf, echo will not necessarily expand escapes.

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -37,12 +37,12 @@ prplmesh_platform_db_init() {
     operating_mode=${2-Gateway}
 
-    mkdir -p @PLATFORM_DB_PATH_TEMP@/
+    mkdir -p /tmp/
```
Rebase screw-up?

tomereli replied:
no, fixing another issue. I didn't know you will be reviewing, I should have opened a draft PR. But its a good thing you review this now

arnout commented on .travis.yml:
```diff
@@ -10,6 +10,10 @@ script:
   - sudo ./image-build.sh -b ubuntu:18.04
   - sudo ./build.sh map -f MSGLIB=zmq BUILD_TESTS=ON
+  # Start controller & agent node detached and check status
+  - sudo ./run.sh start-controller-agent -d -n controller
```
Since it has controller and agent, it would be better to use `-n controller-agent`.

arnout commented on tools/docker/test.sh:
```diff
@@ -0,0 +1,15 @@
+topdir="${scriptdir%/*/*/*}"
+
+main() {
+    run docker container exec -it ${1-agent} ${topdir}/build/install/scripts/prplmesh_utils.sh status
```
It would be better to use the -n option for the name, to make it similar to the run.sh.

arnout commented on .travis.yml:
```diff
@@ -8,8 +8,14 @@ services:
 script:
   - cd tools/docker
-  - sudo ./image-build.sh -b ubuntu:18.04
+  - sudo ./image-build.sh -b ubuntu:19.04
```
Why 19.04?

tomereli replied:
it stil doesn't work, the agent crashes, i'm trying anything I can think of - it certainly not related to our build.. any idea?

tomereli replied:
wasn't related, it appears that our RELEASE is broken on docker container (worked fine locally for me) - see #93

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -8,7 +8,11 @@ dbg() {
 
 err() {
-	echo -e '\033[1;31m'"$@"'\033[0m'
+    printf '\033[1;31m'"$@\n"'\033[0m'
```
I would have put the newline after the end-of-color, but it doesn't matter much. :ok:

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -130,17 +134,49 @@ stop_function() {
+    [ "$error" = "true" ] && cat $LOGS_PATH/beerocks_agent.log
+
+    report "wlan0 radio agent operational" radio_agent_operational $LOGS_PATH wlan0
+    [ "$error" = "true" ] && cat $LOGS_PATH/beerocks_agent_wlan0.log
```
This will dump the wlan0 log if the first check failed... Which maybe you want, but it is a bit weird. :ok:

tomereli replied:
no it isn't, thanks for that

arnout commented on tools/docker/image-build.sh:
```diff
@@ -21,27 +21,31 @@ scriptdir="$(cd "${0%/*}"; pwd)"
     echo "      -v|--verbose - verbosity on"
-    echo "      -b|--base-image - Base OS image to use (Dockerfile 'FROM') "
+    echo "      -b|--base-image - Base OS image to use (Dockerfile 'FROM')"
+    echo "      -B|--builder-name - name for the builder image (default prplmesh-builder)"
```
What is the use case for setting these names?

I would tend to prefer to use a tag instead, and always use the same tag for both. So `-t <tag>` as the option for the script, and `TAG=":$2"` with TAG defaulting to empty, and `--tag=prplmesh-builder$TAG` in the build call.

arnout commented on tools/docker/run.sh:
```diff
@@ -52,18 +53,19 @@ main() {
-    docker image inspect prplmesh-runner >/dev/null 2>&1 || {
-        dbg "Image prplmesh-runner does not exist, creating..."
+    docker image inspect ${IMAGE} >/dev/null 2>&1 || {
+        dbg "Image ${IMAGE} does not exist, creating..."
```
You'll also have to pass the name to the image-build script, of course.

arnout commented on tools/docker/run.sh:
```diff
@@ -52,18 +53,19 @@ main() {
+            -d | --detach)      DETACH=true; shift ;;
+            -i | --ipaddr)      IPADDR="$2"; shift; shift ;;
+            -n | --name)        NAME="$2"; shift; shift ;;
+            -I | --image)  IMAGE="$2"; shift; shift ;;
```
This should be tag as well if we decide to go that way, obviously.

arnout reviewed (approved):
Non-essential changes only, or things that can be done later.

Commit:
CI: update .travis.yaml for operational test

Update .travis.yaml to run operational test during continuous
integration - running 2 nodes, each in a separate container - controller
and agent node, and an agent only node, connected through a docker
bridge.
Testing is using prplmesh_utils.sh status command which checks that all
processes are running, and that both the main agent and radio agents are
operational.
For the agent only node, there is still an issue so we don't fail the
test on this.

Note that we build in debug mode since release build is broken in
containers - see issue
https://github.com/prplfoundation/prplMesh/issues/93

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
common: scripts: prplmesh_utils.sh status

Add operational status check to status sub-command which greps the logs
of the main and radio agents, and output the status to the console.
In addition, change the return value to match the operational status, so
it can be used by CI.

Signed-off-by: Ostap <slyusar.ostap@gmail.com>
Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tools: docker: add test.sh script

Add test.sh, simple wrapper for prplmesh_utils.sh status which will be
called by CI.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tools: docker: runner - fix entrypoint.sh

Fix entrypoint.sh which tried to up the interface while adding it,
causing several OS distros to fail.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
common: build: add BEEROCKS_TMP_PATH

Add CMAKE set for BEEROCKS_TMP_PATH, which will be used by
prplmesh_utils.sh to check logs status.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tools: docker: add option for image tag

Add -t | --tag option for image-build.sh, and make the scripts POSIX
compatible.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
